packages/mongodb-dialects/spring-@query/src/main/kotlin/com/mongodb/jbplugin/dialects/springquery/SpringAtQueryDialectParser.kt

```diff
@@ -0,0 +1,333 @@
+        return findParentMethodWithQueryAnnotation(source) != null
+    }
+
+    override fun attachment(source: PsiElement): PsiElement {
+        return findParentMethodWithQueryAnnotation(source)!!
+    }
+
+    override fun parse(source: PsiElement): Node<PsiElement> {
+        if (source !is PsiMethod) {
+            return Node(source, emptyList())
+        }
+
+        val collection = resolveMethodCollection(source)
+
+        val queryAnnotation = source.annotations.find {
+            it.hasQualifiedName(QUERY_FQN)
+        } ?: return Node(source, listOf(collection))
+
+        val operation = inferCommandFromQuery(source, queryAnnotation)
+        val injectedQueryHost = findInjectedQuery(queryAnnotation)
+        val injectedQuery =
+            injectedQueryHost?.children?.firstOrNull()
+                ?: return Node(source, listOf(operation, collection))
+
+        return Node(
+            source,
+            listOf(
+                operation,
+                collection,
+                HasFilter(recursivelyParseJsonFilter(injectedQuery, source))
+            )
+        )
+    }
+
+    private fun recursivelyParseJsonFilter(source: PsiElement, parent: PsiMethod): List<Node<PsiElement>> {
+        return when (source.elementType?.toString()) {
+            "ARRAY", "OBJECT" -> {
+                return source.children.flatMap {
+                    recursivelyParseJsonFilter(it, parent)
+                }
+            }
+            "PROPERTY" -> {
+                val fieldText = source.children[0].text
+                val operator = Name.from(fieldText.substring(1))
+                if (operator != Name.UNKNOWN &&
+                    operator != Name.IN &&
+                    operator != Name.NIN &&
+                    source.children[1].elementType.toString() == "ARRAY"
+                ) { // it's an operator with children
+                    val opName = Named(Name.from(fieldText.substring(1)))
+                    return listOf(
+                        Node(
+                            source,
+                            listOf(
+                                opName,
+                                HasFilter(
+                                    recursivelyParseJsonFilter(source.children[1], parent)
+                                )
+                            )
+                        )
+                    )
+                } else if (operator != Name.UNKNOWN) { // it's an operator with a single value
+                    val opName = Named(Name.from(fieldText.substring(1)))
+                    return listOf(
+                        Node(
+                            source,
+                            listOf(
+                                opName,
+                                resolveValueReference(source.children[1], parent)
+                            )
+                        )
+                    )
+                } else if (isImplicitAnd(source.children[1])) {
```

Comment:
I feel like we should be a bit cautious when passing indexed values because they might not exist in some cases, as we saw in the past. Unless we are not parsing when the query is incomplete?